dvc/repo/metrics/show.py

```diff
@@ -188,9 +188,15 @@ def _collect_metrics(repo, path, recursive, typ, xpath, branch):
         if not o.metric:
             continue
 
-        # NOTE this case assumes that metric has been modified with
-        # `dvc metric modify -t {} -x {}`, so stage file should contain
-        # info what type metric is and what is its xpath
+        # NOTE this case assumes:
+        # - that metric has been modified with `dvc metric modify -t {} -x {
+        # }` and stage file should contain info what type metric is and what
+        # is its xpath
+        #
+        # or
```

Comment:
it's `and` I believe. type must not be provided via CLI _and_ there should be a metric definition in the DVC-file

if type is provided via CLI it overrides everything, including xpath.

Also, I would check when we calculate type based on the file extension. I hope it is happening only after this code, down the stack.

Reply:
Yes, it is happening inside `_read_metrics` which is happening righ after `_collect_metrics`.